### PR TITLE
[examples/demo] fix demo could not stat up

### LINKED ISSUE #16090: Demo example not working on Mac M1 Pro
### Component(s)

Examples/demo

## Description
Demo example not working on Mac M1 Pro. Specifically demo-server and demo-client are not starting on Mac M1 Pro.

## Steps to Reproduce

1. Clone the repo and go to examples/demo folder
2. Run `docker compose up -d` to launch the demo app

## Expected Result
demo-server and demo-client docker containers start successfully

## Actual Result
demo-server and demo-client docker containers fail to start

Screen Shot 2022-11-05 at 3 58 32 PM

image
image

Both of them display error message `exec /go/bin/main: no such file or directory` in their respective docker container logs

### Collector version

latest

## Environment
Mac M1 Pro

### OpenTelemetry Collector configuration

```yaml
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  prometheus:
    endpoint: "0.0.0.0:8889"
    const_labels:
      label1: value1

  logging:

  zipkin:
    endpoint: "http://zipkin-all-in-one:9411/api/v2/spans"
    format: proto

  jaeger:
    endpoint: jaeger-all-in-one:14250
    tls:
      insecure: true

processors:
  batch:

extensions:
  health_check:
  pprof:
    endpoint: :1888
  zpages:
    endpoint: :55679

service:
  extensions: [pprof, zpages, health_check]
  pipelines:
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging, zipkin, jaeger]
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging, prometheus]
```

### Log output

```shell
exec /go/bin/main: no such file or directory
```

### Additional context

The issue is Mac M1 Pro specific  (arm64) and does not occur on prior Mac (x86_64). Also, i found the issue started occuring after https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/15999 which reduced image sizes was merged. I checked out a commit prior to that and did not encounter the above issue.

### PR DESCRIPTION
**Pull request by JaredTan95**
Signed-off-by: Jared Tan <jian.tan@daocloud.io>

**Description:** <Describe what has changed.>

fix https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16090

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by PassionateDeveloper86** [comment]
This sadly doesnt fix the bug on my machine (W11, Docker for Windows), same error when I apply your fix manuelly and rebuild the images

**Comment by JaredTan95** [comment]
> This sadly doesnt fix the bug on my machine (W11, Docker for Windows), same error when I apply your fix manuelly and rebuild the images

It doesn't make sense, I tested in my local dev.

**Comment by JaredTan95** [comment]
@PassionateDeveloper86 I update dockerfile, could you test it with apple M1 ?

**Comment by saurabhdes** [comment]
Thankyou @JaredTan95 for the PR. I will check it on my Mac M1 Pro later today and post the result here.

**Comment by saurabhdes** [comment]
@JaredTan95 , I checked out your fork and branch - good news is it's now working on my Mac M1 Pro, but with below caveats:
1. 1st time i ran `docker compose up -d` , demo-server started successfully but demo-client failed with error `2022/11/08 17:53:35 Failed to create the collector trace exporter: context deadline exceeded`
2. The i ran `docker compose up -d` again, after that demo-client started sucessfully. 

I can reproduce this consistently, by running 

1. `docker-compose down`
2. `docker-compose up -d` , notice error in demo-client
3. `docker-compose up -d`, now demo-client runs successfully

Attached some screenshots - the 1st and 2nd show the error, the 3rd and 4th show the resolution after running Step 3 above. 
Screenshot 2022-11-08 at 9 47 19 AM
Screenshot 2022-11-08 at 9 50 39 AM
creenshot 2022-11-08 at 9 48 02 AM" src="https://user-images.githubusercontent.com/47896520/200640363-5b1e5cc8-a33a-4974-a277-99655b2dedd2.png">
rcontent.com/47896520/200640321-deec8b7b-6317-4b8c-a9f4-6b1d7bb9eb12.png">

**Comment by JaredTan95** [comment]
yes, it's another issue,because of otel collector not ready before demo-client start up.

**Comment by saurabhdes** [comment]
Yes, i agree it's a separate issue. I found a workaround which is to have `restart: always` in demo-client's docker compose specification, but we may need some time to figure out a  better alternative.

**Comment by JaredTan95** [comment]
> Yes, i agree it's a separate issue. I found a workaround which is to have `restart: always` in demo-client's docker compose specification, but we may need some time to figure out a better alternative.

a better solution I think that add `otel healthcheck` for `demo-client` in docker-compose.

**Comment by saurabhdes** [comment]
thanks @JaredTan95 , checked out latest commit on my M1 Pro and all containers are starting correctly. PR looks good from my side so i am approving.

**Comment by JaredTan95** [comment]
> thanks @JaredTan95 , checked out latest commit on my M1 Pro and all containers are starting correctly. PR looks good from my side .

thx for your testing. 👍

**Review by andrzej-stencel — APPROVED** [review]

File: examples/demo/client/Dockerfile [hunk 1]
```diff
@@ -12,10 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 FROM golang:1.18 as build
-COPY . /usr/src/client/
-WORKDIR /usr/src/client/
+WORKDIR /app/
+COPY . .
 RUN go env -w GOPROXY=direct
-RUN go install ./main.go
+RUN CGO_ENABLED=0 GOOS=linux go build -o main main.go
```
[the remark this answers is not in the document]
Comment (astencel-sumo) [reply examples/demo/client/Dockerfile:18]: The `GOOS=linux` flag does not hurt, but I also don't think it's needed in the Dockerfiles - these Dockerfiles are Linux-specific anyway. There's no version of alpine image for Windows :wink:

[the remark this answers is not in the document]
Comment (mx-psi) [reply examples/demo/client/Dockerfile:18]: > There's no version of alpine image for Windows wink

Yeah, I was just wondering if the first stage was multiarch and Docker was doing something fancy there :shrug: I would rather remove it if it's not necessary

[the remark this answers is not in the document]
Comment (JaredTan95) [reply examples/demo/client/Dockerfile:18]: yes, it's not necessary,we can remove it. I will update it soon.

**Comment by JaredTan95** [comment]
@mx-psi pls help review this :-P

**Review by mx-psi — APPROVED** [review]

**Pull request merged**